app/controllers/agents_controller.rb

```diff
@@ -220,9 +220,9 @@ def redirect_back(message, options = {})
   end
 
   def build_agent
-    @agent = Agent.build_for_type(params[:agent].delete(:type),
+    @agent = Agent.build_for_type(agent_params.delete(:type),
```

Comment:
Your code makes sense to me, and should have the same behavior as before.  It's this existing usage that I find confusing since it deletes in one place and uses that side effect in another.